### ISSUE TITLE
Max value issue

### ISSUE THREAD

**Issue by ceraolo** [OpenModelica/OpenModelica#12495]
### Description
Consider the MWE:
```
package FindMax
  model Test
    TimeMax timeMax annotation (
      Placement(transformation(origin = {14, 0}, extent = {{-10, -10}, {10, 10}})));
    Modelica.Blocks.Sources.RealExpression signal(y = time*sin(time)) annotation (
      Placement(transformation(origin = {-25, 0}, extent = {{-13, -10}, {13, 10}})));
  equation 
    connect(signal.y, timeMax.u) annotation (
      Line(points={{-10.7,0},{2,0}},    color = {0, 0, 127}));
    annotation (
      experiment(StartTime = 0, StopTime = 20, Tolerance = 1e-06, Interval = 0.04));
  end Test;

  block TimeMax 
    extends Modelica.Blocks.Interfaces.SISO;
  equation 
    y = max(u, y);
  end TimeMax;
end FindMax;


```
In Dymola I get the following:

![image](https://github.com/OpenModelica/OpenModelica/assets/7373030/85ed20cd-a4a7-42ea-8250-a51e3d6cd7da)

With OM 1.24.0-dev 101 I get: 
![1a](https://github.com/OpenModelica/OpenModelica/assets/7373030/73549cf5-f7dd-48a0-860c-598b0405d615)

My expected plots are those from Dymola. Am I right?

### Expected Behavior
The output is the maximum over time of the input

**Comment by ceraolo** [comment]
Further addition. 
If I substitute TimeMax with the following:

```
block TimeMax2 
  extends Modelica.Blocks.Interfaces.SISO;
equation
  if u>y then
    der(y) = max(0, der(u));
  else
    der(y)=0;
  end if;
end TimeMax2;
```
the simulation gives the correct result, but we get (for a 20 sec simulation time) 38321 messages of the type 

`Homotopy solver total pivot: Matrix (nearly) singular at time 0.001756.
`

And the simulation becomes slow (0.3 s is much for a sytem having just TimeMax2).

**Comment by casella** [comment]
@ceraolo your TimeMax block is not a well-posed mathematical problem. The equation
```modelica
y = max(u, y);
```
is implicit if solved for y. It can be solved by Newton's method, trying to get the following residual to zero:
```modelica
res = y - if u > y then u else y; 
jac = 1 - if u > y then 0 else 1; 
```
At all points where you expect `y = u`, the residual is theoretically zero, but in practice it will always be slightly positive or negative due to small numerical errors:
- if y > u then res = y - y = 0 (exact zero) and jac = 1 - 1 = 0 (exact zero)
- if y < u then res = y - u = -eps and jac = 1

So, half of the time the Newton update will be 0/0, i.e. it will be undetermined. BTW, this is also what Dymola does:
```
  // Nonlinear system of equations
  // Tag: simulation.nonlinear[1]
    // It depends on the following timevarying variables: 
    //   timeMax.u
    // Start values for iteration variables of non-linear system of 1 equations: 
    //   timeMax.y(start = 0.0)

  equation // Residual equations
    0 = timeMax.y-max(timeMax.y, timeMax.u);
    // Non-zero elements of Jacobian
      J[1, 1] := 1.0-(if noEvent(timeMax.y > timeMax.u) then 1.0 else 0.0);
  // End of nonlinear system of equations
```
I understand Dymola tries to go the extra mile to solve such undetermined systems, I myself am not fully convinced that this is a good idea, because it requires some additional assumptions that are not available to the modeller - IMHO, declarative problems should always be well-posed (in the sense of Hadamard). See also #12357.

What do you think?

**Comment by casella** [comment]
Big deal, see discussion in modelica/ModelicaSpecification#109 and https://stackoverflow.com/questions/30732774/modelica-compute-minimum-maximum-of-continuous-variable-over-time.

I also tried the solution proposed by Marco Bonvini, i.e. replace y with T*der(y) + y

```modelica
package FindMax
  model Test
    TimeMax timeMax annotation (
      Placement(transformation(origin = {14, 0}, extent = {{-10, -10}, {10, 10}})));
    Modelica.Blocks.Sources.RealExpression signal(y = time*sin(time)) annotation (
      Placement(transformation(origin = {-25, 0}, extent = {{-13, -10}, {13, 10}})));
  equation 
    connect(signal.y, timeMax.u) annotation (
      Line(points={{-10.7,0},{2,0}},    color = {0, 0, 127}));
    annotation (
      experiment(StartTime = 0, StopTime = 20, Tolerance = 1e-06, Interval = 0.04));
  end Test;

  block TimeMax 
    extends Modelica.Blocks.Interfaces.SISO;
    parameter Real T = 1e-3;
    Real r;
    Real yd;
  initial equation
    yd = u;
  equation 
    r = max(y,u);
    T*der(yd) + yd = r;
    y = yd;
  end TimeMax;
end FindMax;
```
but it doesn't really work with implicit solvers like DASSL:
![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/445ec5b0-4c7c-42c9-95ab-3fbb012530ce)

Nor it works in Dymola
![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/58ffc186-f08f-46e8-ada7-c171d84a6c8a)

There are also some posts on the web that suggest using `pre(y)`, but the problem is, according to the [MLS Sect. 3.7.5 operator 3.24](https://specification.modelica.org/maint/3.6/operators-and-expressions.html#event-related-operators-with-function-syntax), you can only apply `pre()` to _discrete time expressions_.

I guess this is really a tough problem for a declarative modelling language.

**Comment by casella** [comment]
BTW, what is really weird in the above-mentioned solution is that y and der(y) are inconsistent: in some time intervals the derivative is positive (as it should), but the corresponding variable remains constant. Go figure. I guess that is due to something going very wrong with DASSL in this problem.

**Comment by casella** [comment]
> Only Dymola with my original TimeMax makes the miracle of running without numerical issues, but if this is done by some internal trick, we might question whether this model continues to work well when inserted in larger ones.

Absolutely. We shouldn't rely on magic in mathematical modelling 😅

**Comment by casella** [comment]
Please try it out in some applications, if it works fine, we may add it to the MSL.

**Comment by casella** [comment]
The only problem is that u may not be differentiable. In that case, optionally adding a low-pass filter to the input
```
T*der(uf) + uf = u:
```
and then using `der(uf)` would work, with some approximation depending on the value of the time constant `T`.

**Comment by ceraolo** [comment]
> 
> and then using `der(uf)` would work, 
Yeah, a first-order filter is a friend that solves several situations, from algebraic loops to render derivatives  continuous

**Comment by ceraolo** [comment]
> Please try it out in some applications, if it works fine, we may add it to the MSL.

Tried, but got a strange reaction from MSL official:
https://github.com/modelica/ModelicaStandardLibrary/issues/4414

I would gladly share the proposal with you, maybe implementing the optional filter you suggested 

In the meanwhile, members of my team are starting using TimeMax in some production models

**Comment by casella** [comment]
I'm sorry, I was actually aware that @AHaumer implemented a solution to this problem in the not-so-distant past, but did look for it. The Modelica.Blocks.Math.ContinuousSignalExtrema is definitely interesting. It is currently available in MSL 4.1.0 beta 1, which we will hopefully release in September. You can find the code [here](https://specification.modelica.org/maint/3.6/overloaded-operators.html), try it out and compare with your approach in actual use cases.

The topic is tricky, different implementations may be more or less efficient in different use cases, we need some actual experimental results on that. If you team members can try both and report, that would be great.
